uamqp/message.py

```diff
@@ -116,6 +120,42 @@ def __init__(self,
             self._annotations = annotations
             self._header = header
 
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state["_body"] = list(state["_body"].data) if state["_body"].data else ""
+
+        state["state"] = None
+        state["delivery_no"] = None
+        state["delivery_tag"] = None
+        # need _response for settled property
+        state["_response"] = self.settled
+        state["_settler"] = None
+        state["on_send_complete"] = None
+        state["_message"] = None
```

Comment:
I'm wondering setting them to None is not enough, or it should be `del state[key]`.

Do you have an sample of the pickled bytes of a message object?

This also make me think that we probably should echo what we did for MessageHeader/MessageProperties: creating an empty dic and then pick the properties that we care.

Reply:
So I decided to set the `state[key] = None` rather than deleting them b/c, for instance, if a user tries to pickle an event and repickle it, deleting the key results in a `KeyError`. for example:
```
 pickle_ed = pickle.loads(pickle.dumps(ed))
 repickle_ed = pickle.loads(pickle.dumps(pickle_ed))   # results in KeyError
```
When you say sample of the pickled bytes, do you mean maybe a pickle file of a pickled message object?

Makes sense. I believe the properties that would require getstate/setstate for their respective types are _message, _settler, on_send_complete, and state. The `cMessage` class for `_message` is the only one I don't think I should touch, since it deals with the c files??

Reply:
based on our discussion, in __getstate__ we pick a minimum set of objects to serialize:
- delivery_no
- delivery_tag
- properties
- application_properties 
- annotations
- header
- footer
- delivery_annotations
- _body
- _body_type
(feel free to call out if there's any other private objects that are required to be encoded)
so we would be doing:

```python
def __getstate__(self):
   state = {}
   state["application_properties"] = self.application_properties
   ...
   return state
```

as far as I could see, this gives us two benefits:

1. extensibility, in the future we might want to add/remove instance variables
2. save encoded bytes: useless objects will not be encoded to save bytes

Reply:
and in the __setstate__, we manually populate the instance variables by default value which are not encoded:
```python
    def __setstate__(self, state):
        state["state"] = constants.MessageState.WaitingToBeSent
        state["idle_time"] = ...
        ...
        self.__dict__.update(state)
        # populate body
```

Reply:
that makes sense to me! One more property that is relevant is `settled`, which is determined by `_response`. To only expose `settled`, I plan to do something like the following:
```
def __getstate__(self):
    ....
    state['settled'] = self.settled  # True/False
    ...
def __setstate__(self, state):
    state['_response'] = errors.MessageAccepted()  if state['settled'] else None # arbitrary value for settled to be True

Reply:
I don't think we could simply set state['_response'] to errors.MessageAccepted() based on `settled` is True or not.
because self._response could be
`MessageAlreadySettled`
`MessageAccepted`
`MessageRejected`
`MessageReleased`
`MessageModified`

depending on how `self._response` is used in current message (mostly for checking whether a message could be settled), then the public settled property is sufficient.

so I would keep `state['_response'] = None` for now to avoid introducing complexity.

Reply:
since `def settled` method doesn't have a corresponding setter method, and it relies on `_response` for a return value, wouldn't we need to also serialize `_response`? If we set it to none, then even if `message.settled` was True before pickling, it would be False after.